acra-migrate-keys tool

ilammy opened the pull request:
Create a tool for migrating keys between key store formats. Currently it supports only v1 ⟹ v2 migration.

With default settings it is enough to simply run `acra-migrate-keys` to migrate keys from `.acrakeys` (v1) into `.acrakeys.migrated` (v2). After successful migration the administrator has to manually reconfigure Acra services to use the new key store. It is possible to explicitly specify migration details via `--src...` and `--dst...`  options if necessary.

There is a `--dry_run` option which performs migration in memory, without writing results to disk. You can use it to test that everything is in place, the keys can be read and understood.

By default the migration tool will refuse to write output into already existing key store. This is because the migration process is not transactional and idempotent. Migrating the same key store twice will result in duplicate keys, and failure to migrate some keys will still migrate other successful keys. This behavior can be overridden with `--force` option, allowing the migration to proceed.

Note that master keys are expected in environment variables, as usual. You will need to set all variables for successful migration:

  - ACRA_MASTER_KEY
  - ACRA_MASTER_ENCRYPTION_KEY
  - ACRA_MASTER_SIGNATURE_KEY

More implementation details are in commit messages. The code is structured in a way that should easy its reuse in Acra EE with its extended key store support (both v1 and v2).

Lagovas commented on keystore/v2/keystore/importV1.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func zeroizeSymmetricKey(key []byte) {
+	utils.FillSlice(0, key)
```
sometimes we will decide to refactor FillSlice to something like that )) or we will not only zeroize but oneize, secondize too ))

ilammy replied:
Yeah, now that you've mentioned it, this definitely looks like something for `utils`. Should we move these functions there? I remember you've added some in #355:
https://github.com/cossacklabs/acra/pull/355/files#diff-45a2ca38910926a14a2c14c6a04192dd

Lagovas replied:
I think will be nice to avoid duplication if sometimes we need it one more time

ilammy replied:
I’ve moved the utilities to `utils`. They will be available for general use once **keystore-v2** lands into master.

Lagovas commented on tests/test.py:
```diff
@@ -4253,6 +4255,7 @@ def testStartupWithoutVersionInConfig(self):
             default_args = {
                 'acra-server': ['-db_host=127.0.0.1'],
                 'acra-connector': ['-user_check_disable', '-acraserver_connection_host=127.0.0.1', '-client_id=keypair1'],
+                'acra-migrate-keys': ['--dry_run', '--src_keydir={}'.format(KEYS_FOLDER.name)],
```
it's cool that you found these place in our 4.4k lines file with tests where you can test starting tool ))

ilammy replied:
Well, no silver bullet here, I just followed the stack trace:

    Traceback (most recent call last):
      File "tests/test.py", line 4282, in testStartupWithOutdatedConfigVersion
        stderr = self.getOutputFromProcess(args)
      File "tests/test.py", line 1224, in getOutputFromProcess
        process = subprocess.Popen(args, stderr=subprocess.PIPE)
      File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
        restore_signals, start_new_session)
      File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
        raise child_exception_type(errno_num, err_msg)
    FileNotFoundError: [Errno 2] No such file or directory: './acra-migrate-keys'

If not for that, I'd still be confused about what has gone wrong. So, ugh, thanks for using Python, I guess :)

ilammy commented on keystore/filesystem/key_export.go:
```diff
@@ -0,0 +1,298 @@
+// EnumerateExportedKeys prepares a list of keys that can be exported.
+// The keys are classified with default key file classifier.
+func (store *KeyStore) EnumerateExportedKeys() ([]ExportedKey, error) {
+	return store.EnumerateExportedKeysByClass(&DefaultKeyFileClassifier{})
```
Mmm... It seems like we wouldn't need any state here, and there will be at most two implementations, for CE and EE respectively. However, I'm not sure what do mean with package variables. Could you please elaborate?

I've intended for user code (i.e., migration tool) to only call `EnumerateExportedKeys()`. This other method `EnumerateExportedKeysByClass()` will be used by Acra EE code to implement `EnumerateExportedKeys()` for its basic key store, passing an extended classifier. Migration tool in Acra EE will still use `EnumerateExportedKeys()` (this time for EE key store).

Lagovas replied:
to avoid each time allocation of empty struct (I dont know how golang actually compile it, but can predict that at least some pointer it may allocate on stack/heap), we can allocate it once on package level and use it every time when need (like singleton). For example how it used in flag package - https://golang.org/src/flag/flag.go#L1010
```
var defaultClassifier = &DefaultKeyFileClassifier{};

func (store *KeyStore) EnumerateExportedKeys() ([]ExportedKey, error) {
	return store.EnumerateExportedKeysByClass(defaultClassifier)
}
```

ilammy replied:
Oh, I’ve misunderstood your original comment. I thought you meant to use global variables as a way to configure the “classifier” being used by default, with the idea that Acra EE will somehow set a new value for this value, substituting its own classifier.

But you seem to have meant the allocation optimization as above. I’ve updated the code to include it like you’ve written.

ilammy commented:
The dependencies have been merged, but the build has been broken. I guess I'll fix it up first (should require only a few tweaks, I hope), and then rebase and retarget this PR onto the correct branch before undrafting it.

ilammy commented on cmd/acra-migrate-keys/acra-migrate-keys.go:
```diff
@@ -0,0 +1,245 @@
+	log.Trace("enumerating keys for export")
+	keys, err := srcV1.EnumerateExportedKeys()
+	if err != nil {
+		log.WithError(err).Debug("failed to enumerate exported keys")
```
> But `errors.New` should be defined with lowercase message according to golang standard libraries.

Oh, I just noticed that I've been writing errors like this:


```go
var (
        ErrUnknownPurpose = errors.New("KeyStore: unknown key purpose")
)
```

with a `KeyStore:` prefix.

I guess it is redundant? Internal key store loggers have a `service=keystore` field added to them so we keep this context. If these errors bubble up somewhere else they seem to be not too out-of-context to not require these prefixes.

I think I'll cook a PR for that...

ilammy commented:
Rebased and retargeted. Changes since draft:

- Updated log lines to start with an uppercase letter to conform with the style.
- Slightly renamed the src/dst options to be consistent with other tools.
- Slightly improved handling of command-line options to avoid dereferences everywhere.
- Added utilities for exporting both encrypted and plaintext symmetric keys (useful for EE).
- Unbroken the build ;)

Lagovas commented:
lgtm

ilammy commented:
Hm... Filesystem key store tests just failed on this push, there was a problem in locking tests (should not be affected by these changes at all). I’ve restarted the build to see if it persists.

I’ll keep it mind, it might be possible that the locking implementation is broken in a subtle way.

vixentael commented on cmd/acra-migrate-keys/acra-migrate-keys.go:
```diff
@@ -0,0 +1,245 @@
+func RegisterCommandLineParams() *CommandLineParams {
+	params := &CommandLineParams{}
+	// Source key store
+	flag.StringVar(&params.Src.KeyStoreVersion, "src_keystore", "v1", "key store format to use: v1 (current), v2 (experimental)")
```
I'd not name v2 as `experimental`, because when this thing is merged to master, we'll totally forget to rename these into "old" and "new" (or similar)
